### PR TITLE
feat: Integrate Flowkana into ERPNext

### PR DESCRIPTION
**Pull request by rvpasari**
https://app.asana.com/0/1192407975826276/1199722767556050

Covers subtask 5,6 and 7

### CODE REVIEW

File: erpnext/erpnext_integrations/doctype/flowkana_settings/flowkana_settings.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+			doc.enable = 1
+		else:
+			doc.enable = 0
+		doc.save()
+
+	def check_fulfillment_partner(self):
+		"""
+		Create a fulfillment partner in case they don't exist.
+		"""
+		if not frappe.get_doc("Fulfillment Partner", "Flowkana"):
```
Comment (sahil28297) [inline erpnext/erpnext_integrations/doctype/flowkana_settings/flowkana_settings.py]:
```suggestion
		if not frappe.db.exists("Fulfillment Partner", "Flowkana"):
```

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 2]
```diff
@@ -189,6 +191,13 @@ def on_submit(self):
 			from erpnext.accounts.doctype.pricing_rule.utils import update_coupon_code_count
 			update_coupon_code_count(self.coupon_code,'used')
 
+		flowkana_settings = frappe.get_doc("Flowkana Settings")
+
+		#flowkana
+		if flowkana_settings.enable_flowkana and self.fulfillment_partner == "Flowkana":
+			print("sending request to flowkana endpoint....")
```
Comment (sahil28297) [inline erpnext/selling/doctype/sales_order/sales_order.py]:
```suggestion
```

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 3]
```diff
@@ -475,6 +484,86 @@ def validate_serial_no_based_delivery(self):
...
+		integration_request = frappe.new_doc("Integration Request")
+		integration_request.update({
+			"integration_type": "Host",
+			"integration_request_service": "Flowkana",
+			"status": "Queued",
+			"data": json.dumps(request_json, default=json_handler),
+			"reference_doctype": "Sales Order",
+			"reference_docname": self.name
+		})
+		integration_request.insert()
```
Comment (sahil28297) [inline erpnext/selling/doctype/sales_order/sales_order.py]:
```suggestion
		integration_request.save(ignore_permissions=True)
```

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 4]
```diff
@@ -475,6 +484,86 @@ def validate_serial_no_based_delivery(self):
...
+						"customer_license": self.license,
+						"delivery_date": self.delivery_date,
+						"note": "Test Note"
+					},
+					"relationships": {
+						"order_line_items": item_list
+					}
+				}
+			}
+		print("Request JSON: ", request_json)
```
Comment (sahil28297) [inline erpnext/selling/doctype/sales_order/sales_order.py]:
```suggestion
```

**Review by sahil28297 — DISMISSED** [review]

File: erpnext/selling/doctype/sales_order/sales_order.py (hunk 3 above)
Comment (rvpasari) [reply erpnext/selling/doctype/sales_order/sales_order.py]: We will need to insert it first

File: erpnext/erpnext_integrations/doctype/flowkana_settings/flowkana_settings.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+		self.check_fulfillment_partner()
+		self.toggle_status()
+
+	def toggle_status(self):
+		doc = frappe.get_doc("Fulfillment Partner", "Flowkana")
+		if self.enable_flowkana:
+			doc.enable = 1
+		else:
+			doc.enable = 0
+		doc.save()
```
Comment (hrwX) [inline erpnext/erpnext_integrations/doctype/flowkana_settings/flowkana_settings.py]: - Use frappe.db.set_value

File: erpnext/erpnext_integrations/doctype/flowkana_settings/flowkana_settings.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+	def check_fulfillment_partner(self):
+		"""
+		Create a fulfillment partner in case they don't exist.
+		"""
+		if not frappe.db.exists("Fulfillment Partner", "Flowkana"):
+			if self.enable_flowkana:
```
Comment (hrwX) [inline erpnext/erpnext_integrations/doctype/flowkana_settings/flowkana_settings.py]:
```suggestion
		if self.enable_flowkana and not frappe.db.exists("Fulfillment Partner", "Flowkana"):
```

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 7]
```diff
@@ -189,6 +191,12 @@ def on_submit(self):
 			from erpnext.accounts.doctype.pricing_rule.utils import update_coupon_code_count
 			update_coupon_code_count(self.coupon_code,'used')
 
+		flowkana_settings = frappe.get_doc("Flowkana Settings")
+
+		#send delivery request to flowkana if enabled by user
+		if flowkana_settings.enable_flowkana and self.fulfillment_partner == "Flowkana":
+			self.send_delivery_request_to_flowkana()
+
```
Comment (hrwX) [inline erpnext/selling/doctype/sales_order/sales_order.py]: Use frappe.hooks to integrate it with Flowkana, make sure all the flowkana code is in flowkana's controller

File: erpnext/selling/doctype/sales_order/sales_order.py (hunk 3 above)
Comment (hrwX) [reply erpnext/selling/doctype/sales_order/sales_order.py]: - Save inserts it, if the document is not inserted

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 8]
```diff
@@ -475,6 +483,86 @@ def validate_serial_no_based_delivery(self):
...
+		#mark integration request status as queued, update status to queued
+		integration_request.output = json.dumps(response_data, default=json_handler)
+		integration_request.save(ignore_permissions=True)
+
+
+def json_handler(obj):
+	if isinstance(obj, (datetime.date, datetime.timedelta, datetime.datetime)):
+		return text_type(obj)
+
```
Comment (hrwX) [inline erpnext/selling/doctype/sales_order/sales_order.py]: - Put this in flowkana's controller

**Review by hrwX — DISMISSED** [review]

File: erpnext/erpnext_integrations/doctype/flowkana_settings/flowkana_settings.py [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+	def validate(self):
+		self.check_fulfillment_partner()
+		self.toggle_status()
+
+	def toggle_status(self):
+		doc = frappe.get_doc("Fulfillment Partner", "Flowkana")
+		if self.enable_flowkana:
+			frappe.db.set_value("Fulfillment Partner", "Flowkana", "enable_flowkana", 1)
+		else:
+			frappe.db.set_value("Fulfillment Partner", "Flowkana", "enable_flowkana", 0)
```
Comment (hrwX) [inline erpnext/erpnext_integrations/doctype/flowkana_settings/flowkana_settings.py]:
```suggestion
		frappe.db.set_value("Fulfillment Partner", "Flowkana", "enable_flowkana", self.enable_flowkana)
```

File: erpnext/erpnext_integrations/doctype/flowkana_settings/flowkana_settings.py [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+		"""
+		Create a fulfillment partner in case they don't exist.
+		"""
+		if self.enable_flowkana and not frappe.db.exists("Fulfillment Partner", "Flowkana"):
+			partner = frappe.get_doc({
+				"doctype": "Fulfillment Partner",
+				"partner_name": "Flowkana",
+				"enable": 1
+			})
+			partner.save()
```
Comment (hrwX) [inline erpnext/erpnext_integrations/doctype/flowkana_settings/flowkana_settings.py]:
```suggestion
			partner = frappe.get_doc({
				"doctype": "Fulfillment Partner",
				"partner_name": "Flowkana",
				"enable": 1
			}).save()
```

File: erpnext/erpnext_integrations/doctype/flowkana_settings/flowkana_settings.py [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+					"customer_name": sales_order.customer,
+					"customer_license": sales_order.license,
+					"delivery_date": str(sales_order.delivery_date),
+					"note": "Test Note"
+				},
+				"relationships": {
+					"order_line_items": item_list
+				}
+			}
+		}
```
Comment (hrwX) [inline erpnext/erpnext_integrations/doctype/flowkana_settings/flowkana_settings.py]:
```suggestion
	request_json = {
		"data": {
			"attributes": {
				"external_order_id": sales_order.name,
				"customer_name": sales_order.customer,
				"customer_license": sales_order.license,
				"delivery_date": str(sales_order.delivery_date),
				"note": "Test Note"
			},
			"relationships": {
				"order_line_items": item_list
			}
		}
	}
```

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 12]
```diff
@@ -189,6 +192,13 @@ def on_submit(self):
 			from erpnext.accounts.doctype.pricing_rule.utils import update_coupon_code_count
 			update_coupon_code_count(self.coupon_code,'used')
 
+		flowkana_settings = frappe.get_doc("Flowkana Settings")
+
+		#send delivery request to flowkana if enabled by user
+		if flowkana_settings.enable_flowkana and self.fulfillment_partner == "Flowkana":
+			send_delivery_request_to_flowkana(self.name)
```
Comment (hrwX) [inline erpnext/selling/doctype/sales_order/sales_order.py]:
```suggestion
			send_delivery_request_to_flowkana(self)
```

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 13]
```diff
@@ -189,6 +192,13 @@ def on_submit(self):
 			from erpnext.accounts.doctype.pricing_rule.utils import update_coupon_code_count
 			update_coupon_code_count(self.coupon_code,'used')
 
+		flowkana_settings = frappe.get_doc("Flowkana Settings")
```
Comment (hrwX) [inline erpnext/selling/doctype/sales_order/sales_order.py]: - Cache Flowkana Settings doc

File: erpnext/erpnext_integrations/doctype/flowkana_settings/flowkana_settings.py [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+		"""
+		if self.enable_flowkana and not frappe.db.exists("Fulfillment Partner", "Flowkana"):
+			partner = frappe.get_doc({
+				"doctype": "Fulfillment Partner",
+				"partner_name": "Flowkana",
+				"enable": 1
+			})
+			partner.save()
+
+def send_delivery_request_to_flowkana(sales_order_name):
```
Comment (hrwX) [inline erpnext/erpnext_integrations/doctype/flowkana_settings/flowkana_settings.py]:
```suggestion
def send_delivery_request_to_flowkana(sales_order):
```

File: erpnext/erpnext_integrations/doctype/flowkana_settings/flowkana_settings.py [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+	"""
+	Ping flowkana with sales order details and map response to an integration request.
+
+	Args:
+		sales_order_name (string): name of the sales order that needs to be sent to flowkana
+	"""
+	sales_order = frappe.get_doc("Sales Order", sales_order_name)
```
Comment (hrwX) [inline erpnext/erpnext_integrations/doctype/flowkana_settings/flowkana_settings.py]:
```suggestion
```

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 16]
```diff
@@ -475,6 +485,82 @@ def validate_serial_no_based_delivery(self):
...
+			json=request_json)
+		response_data = response.json()
+
+		if response.status_code not in [200, 201, 202]:
+			frappe.throw(_("The response has the following errors: {0}".format(response_data.get("errors"),"")))
+			return
+
+		#mark integration request status as queued, update status to queued
+		integration_request.output = json.dumps(response_data, default=json_handler)
+		integration_request.save(ignore_permissions=True)
```
Comment (hrwX) [inline erpnext/selling/doctype/sales_order/sales_order.py]: This is already defined in flowkana settings

**Review by hrwX — CHANGES_REQUESTED** [review]

File: erpnext/stock/doctype/item/item.json [hunk 17]
```diff
@@ -1115,14 +1117,24 @@
...
    "label": "Shipping Information",
    "options": "Shipping Information"
+  },
+  {
+   "fieldname": "flowkana_details_section",
+   "fieldtype": "Section Break",
+   "label": "Flowkana Details"
+  },
+  {
+   "fieldname": "ivt_id",
```
Comment (hrwX) [inline erpnext/stock/doctype/item/item.json:1127]:
```suggestion
   "fieldname": "inventory_id",
```

**Review by hrwX — CHANGES_REQUESTED** [review]

File: erpnext/stock/doctype/item/item.json (hunk 17 above)
Comment (rvpasari) [reply erpnext/stock/doctype/item/item.json:1127]: @hrwX Flowkana uses ivt_id and I wanted to keep it consistent - is that okay?

File: erpnext/erpnext_integrations/doctype/flowkana_settings/flowkana_settings.py [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+		"integration_request_service": "Flowkana",
+		"status": "Queued",
+		"data": json.dumps(request_json, default=json_handler),
+		"reference_doctype": "Sales Order",
+		"reference_docname": sales_order.name
+	})
+	integration_request.save(ignore_permissions=True)
+
+	#fetch and prepare headers from flowkana settings, flag error if missing data
+	flowkana_settings = frappe.get_doc("Flowkana Settings")
```
Comment (sahil28297) [inline erpnext/erpnext_integrations/doctype/flowkana_settings/flowkana_settings.py]:
```suggestion
	flowkana_settings = frappe.get_cached_doc("Flowkana Settings")
```

**Review by sahil28297 — CHANGES_REQUESTED** [review]

**Review by sahil28297 — DISMISSED** [review]

**Comment by sahil28297** [comment]
@rvpasari why is this in bs-mar-sprint-1 branch?

File: erpnext/erpnext_integrations/doctype/flowkana_settings/flowkana_settings.py (hunk 18 above)
Comment (sahil28297) [inline erpnext/erpnext_integrations/doctype/flowkana_settings/flowkana_settings.py]:
```suggestion
	flowkana_settings = frappe.get_cached_doc("Flowkana Settings")
```

**Review by sahil28297 — CHANGES_REQUESTED** [review]

**Review by sahil28297 — APPROVED** [review]

**Pull request merged**